### PR TITLE
feat: redirect after login to entrance url

### PR DESCRIPTION
**Pull request by mshannaq**
Redirect after login to entrance url 
closes #722

### CODE REVIEW

**Comment by datamweb** [comment]
Thank you very much for your PR.

Please note that commit writing must be with the following description:
https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/workflow.md#committing

Please write unit test for PR:
https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/pull_request.md#unit-testing

**Comment by kenjis** [comment]
The `develop` branch has been updated and fixed the coding style check errors.
Please rebase to update your PR branch.
See https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/workflow.md#updating-your-branch

**Comment by datamweb** [comment]
@mshannaq  Please do not use `git merge`.
Use `git rebase` to update the branch.
Using  `git merge` caused the github action to fail:
https://github.com/codeigniter4/shield/actions/runs/5891110836/job/15977524828?pr=789

**Comment by datamweb** [comment]
It seems you have 2 ways to fix error **Detect Merge Commits**. 
1. use `git rebase -i`
2. use `git cherry-pick`

I try to use `git rebase -i` to fix the Github Action:
```console
git branch beforeLogginUrl.bk beforeLogginUrl
```
```console
git rebase -i origin/beforeLogginUrl~5
```
After your above command, your editor will open as follows.

Screenshot 2023-08-18 185910

Now modify the items in the editor as follows. And after that save the file and close the editor:

Screenshot 2023-08-18 191009

Now wait a bit. If you are asked for a password, enter your password. Your editor will open again as follows:

Screenshot 2023-08-18 191337

You should edit the file as follows and then save it. After that, close the editor:

Screenshot 2023-08-18 191802

If you have done everything correctly, you should see the following message in your terminal.

Screenshot 2023-08-18 191918

```console
git push -f origin beforeLogginUrl
```
The above method is called squash. After executing the commands, your **5 commits** should become **1 commit** with message `feat: redirect after login to entrance url`.

To be honest, these topics are complicated for me too, I hope you can handle it.

**Comment by kenjis** [comment]
@mshannaq After `git rebase`, there are three commits. But it seems there is no reason to be three commits. The commit messages are the same, and they are all for `redirect after login to entrance url`.

If you can, please squash them into one commit.

File: src/Config/Auth.php [hunk 1]
```diff
@@ -409,7 +409,8 @@ class Auth extends BaseConfig
      */
     public function loginRedirect(): string
     {
-        $url = setting('Auth.redirects')['login'];
+        $session = session();
+        $url     = $session->getTempdata('beforeLogginUrl') ?? setting('Auth.redirects')['login'];
```
Comment (kenjis) [inline src/Config/Auth.php:413]:
```suggestion
        $url     = $session->getTempdata('beforeLoginUrl') ?? setting('Auth.redirects')['login'];
```

File: src/Filters/SessionAuth.php [hunk 2]
```diff
@@ -75,6 +75,11 @@ public function before(RequestInterface $request, $arguments = null)
                 ->with('error', $authenticator->getPendingMessage());
         }
 
+        if (! url_is('login')) {
+            $session = session();
+            $session->setTempdata('beforeLogginUrl', current_url(), 300);
```
Comment (kenjis) [inline src/Filters/SessionAuth.php:80]:
```suggestion
            $session->setTempdata('beforeLoginUrl', current_url(), 300);
```

File: tests/Authentication/Filters/SessionFilterTest.php [hunk 3]
```diff
@@ -83,4 +83,12 @@ public function testBlocksInactiveUsers(): void
 
         setting('Auth.actions', ['register' => null]);
     }
+
+    public function testStoreRedirectsToEntraceUrlIntoSession(): void
+    {
+        $result = $this->call('get', 'protected-route');
+        $result->assertRedirectTo('/login');
+        $this->assertNotEmpty($_SESSION['beforeLogginUrl']);
+        $this->assertSame(site_url('protected-route'), $_SESSION['beforeLogginUrl']);
```
Comment (kenjis) [inline tests/Authentication/Filters/SessionFilterTest.php:92]: All superglobals should be removed. So please use the Session class.
```suggestion

        $result->assertRedirectTo('/login');

        $session = session();
        $this->assertNotEmpty($session->get('beforeLogginUrl'));
        $this->assertSame(site_url('protected-route'), $session->get('beforeLogginUrl'));
```

**Comment by mshannaq** [comment]
I created a new branch and apply the changes into as it more easy and then I submit a PR793

**Pull request closed without merging**